### PR TITLE
Web API for benches and reviews

### PR DESCRIPTION
**Pull request by cvera-tech**
Implemented WebAPI actions for use with the interactive map.

BenchApiController (#9)
- Add a bench
- Get a bench by its ID
- List all benches

ReviewApiController (#24) 
- Add a review to a bench
- Get a review by its ID
- List all reviews
- List all reviews for a bench given its ID

### CODE REVIEW

File: src/AirBench/Api/Controllers/BenchController.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        public async Task<BenchAddResponse> Add(BenchAddRequest request)
+        {
+            var bench = new Bench()
+            {
+                Description = request.Description,
+                Latitude = request.Latitude,
+                Longitude = request.Longitude,
+                NumberSeats = request.NumberSeats
+            };
+            var benchId = await _benchRepo.AddAsync(bench);
```
Comment (appacademy-jchurchill) [inline src/AirBench/Api/Controllers/BenchController.cs:54]: What happens when the HTTP post request doesn't have values for each of the `BenchAddRequest` properties?

File: src/AirBench/App_Start/WebApiConfig.cs [hunk 2]
```diff
@@ -16,7 +13,7 @@ public static void Register(HttpConfiguration config)
 
             config.Routes.MapHttpRoute(
                 name: "DefaultApi",
-                routeTemplate: "api/{controller}/{id}",
+                routeTemplate: "api/{controller}/{action}/{id}",
```
Comment (appacademy-jchurchill) [inline src/AirBench/App_Start/WebApiConfig.cs:16]: Why did you add an `action` route parameter? Are you actually using it? Feels like a violation of REST API design principles.

File: src/AirBench/Data/BenchContext.cs [hunk 3]
```diff
@@ -15,9 +15,14 @@ public BenchContext() : base("bestBenches")
...
+        int IBenchContext.SaveChanges()
         {
-            base.SaveChanges();
+            return base.SaveChanges();
+        }
+
+        async Task<int> IBenchContext.SaveChangesAsync()
+        {
+            return await base.SaveChangesAsync();
         }
```
Comment (appacademy-jchurchill) [inline src/AirBench/Data/BenchContext.cs:26]: Did you need to explicitly define these interface methods on this class? You should be able to add methods in the `DbContext` base class to your `IBenchContext` interface without having to define these passthrough methods.

File: src/AirBench/Api/Repositories/BenchApiRepository.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+﻿using AirBench.Data;
+using AirBench.Models;
+using System.Collections.Generic;
+using System.Data.Entity;
+using System.Threading.Tasks;
+
+namespace AirBench.Api.Repositories
+{
+    public class BenchApiRepository : IBenchApiRepository
```
Comment (appacademy-jchurchill) [inline src/AirBench/Api/Repositories/BenchApiRepository.cs:9]: Was it necessary to add a separate repository for your API controllers?

**Review by appacademy-jchurchill — APPROVED** [review]
Love all of the great documentation that you added to your API controllers. As usual, really strong overall effort John :)

File: src/AirBench/App_Start/WebApiConfig.cs (hunk 2 above)
Comment (cvera-tech) [reply src/AirBench/App_Start/WebApiConfig.cs:16]: I added the action parameter because I wanted to avoid using the RoutePrefix and the Route attributes like Curtis had us do for the DeckOfCards API. I consulted this to design my route actions: 

https://docs.microsoft.com/en-us/aspnet/web-api/overview/web-api-routing-and-actions/routing-in-aspnet-web-api#routing-by-action-name

I will refactor my API to better conform with RESTful principles.

File: src/AirBench/Api/Controllers/ReviewController.cs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        /// <param name="request">The request body.</param>
+        /// <returns>The response body.</returns>
+        [HttpPost]
+        public async Task<ReviewAddResponse> Add(int id, ReviewAddRequest request)
+        {
+            var review = new Review()
+            {
+                BenchId = id,
+                Description = request.Description,
+                Rating = request.Rating
```
Comment (arzonca1) [inline src/AirBench/Api/Controllers/ReviewController.cs:49]: Reviews/comments also require a field for DateTime or DateTimeOffset when the comment was created on, and something to link them to the author since you need to also list the author's name and last name initial at the comment. 

You'd need to likely possibly change that in the database and in the relevant data structures (Review.cs, ReviewAddRequest.cs, etc). Here you could just then do something simple like CreatedOn = DateTimeOffset.now; 

For the author I see that your customprincipal has an Id field and you can probably pass it here so you don't have to query the DB for it.

File: src/AirBench/Api/Controllers/ReviewController.cs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        ///             .
+        ///             .
+        ///             .
+        ///     ]
+        /// }
+        /// </summary>
+        /// <param name="id">The bench ID.</param>
+        /// <returns>The response body.</returns>
+        [HttpGet]
+        public async Task<ReviewListForResponse> ListFor(int id)
```
Comment (arzonca1) [inline src/AirBench/Api/Controllers/ReviewController.cs:172]: I am somewhat confused by creating a separate class just to encapsulate a List<Review> and BranchId, especially when Branch class itself already contains a List<Review> type object.
